### PR TITLE
fix: make PKCS#7 parameter optional as per RFC 5280

### LINKED ISSUE #829: Make PKCS#7 keyEncryptionAlgorithm parameter optional
RFC 5652 - Cryptographic Message Syntax (CMS) section 12.1 defines keyEncryptionAlgorithm ::= algorithmIdentifier to be imported from RFC 5280 Section 4.1.1.2 showing parameter should be an optional field.

https://tools.ietf.org/html/rfc5652#section-12.1
https://tools.ietf.org/html/rfc5280#section-4.1.1.2

Indeed when I try to use `forge.pkcs7.messageFromAsn1(myPKCS7Data)` with no parameter given to the keyEncryptionAlgorithm I get the following error:

```
Error: Cannot read PKCS#7 RecipientInfo. ASN.1 object is not an PKCS#7 RecipientInfo.
    at _recipientFromAsn1 (/redacted/node_modules/node-forge/lib/pkcs7.js:828:17)
    at _recipientsFromAsn1 (/redacted/node_modules/node-forge/lib/pkcs7.js:890:14)
    at Object.fromAsn1 (/redacted/node_modules/node-forge/lib/pkcs7.js:604:24)
    at Object.p7.messageFromAsn1 (/redacted/node_modules/node-forge/lib/pkcs7.js:118:7)
    at pkiOperation (/redacted/redacted.js:62:32)
    at Object.<anonymous> (/redacted/redacted.js:69:1)
    at Module._compile (internal/modules/cjs/loader.js:1185:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1205:10)
    at Module.load (internal/modules/cjs/loader.js:1034:32)
    at Function.Module._load (internal/modules/cjs/loader.js:923:14) {
  errors: [
    '[RecipientInfo.keyEncryptionAlgorithm] Tag class "0", type "16" expected value length "2", got "1"',
    '[RecipientInfo] Tag class "0", type "16" expected value length "4", got "4"'
  ]
}
```
I will submit a pull request momentarily with a fix. Thanks to @NuSkooler for figuring this out and fixing the issue.

### PR DESCRIPTION
**Pull request by troyfactor4**
Fixes #829

### CODE REVIEW

**Review by dlongley — CHANGES_REQUESTED** [review]
Thanks! This PR needs some additional changes down in the `_encryptedContentToAsn1` function to handle when the parameter is `null` or `undefined` (!ec.parameter). That should ensure we can still round trip (parse and reserialize) here.

**Comment by troyfactor4** [comment]
Good catch! Here is my proposed remedy. I've checked asn1.create seems to happily take undefined in an array and remove it. Let me know if you think this is an ok way to handle it.

File: lib/pkcs7.js [hunk 1]
```diff
@@ -837,7 +837,7 @@ function _recipientFromAsn1(obj) {
     serialNumber: forge.util.createBuffer(capture.serial).toHex(),
     encryptedContent: {
       algorithm: asn1.derToOid(capture.encAlgorithm),
-      parameter: capture.encParameter.value,
+      parameter: capture.encParameter ? capture.encParameter.value : null,
```
Comment (dlongley) [inline lib/pkcs7.js:840]:
```suggestion
      parameter: capture.encParameter ? capture.encParameter.value : undefined,
```

File: lib/pkcs7.js [hunk 2]
```diff
@@ -1124,8 +1124,10 @@ function _encryptedContentToAsn1(ec) {
       asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
         asn1.oidToDer(ec.algorithm).getBytes()),
       // Parameters (IV)
-      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
-        ec.parameter.getBytes())
+      ( ec.parameter ?
+        asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
+        ec.parameter.getBytes()) : undefined
+      )
```
Comment (dlongley) [inline lib/pkcs7.js:1130]:
```suggestion
      !ec.parameter ?
        undefined :
        asn1.create(
          asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
          ec.parameter.getBytes())
```

**Review by dlongley — APPROVED** [review]

**Comment by dlongley** [comment]
Thanks!

**Comment by troyfactor4** [comment]
Awesome! Really appreciate the help and fast turn around.

I don't suppose I could beg you to look at #400 when you have 5 min free? Pretty please?

**Review by eiygy — COMMENTED** [review]
- [ ]

**Pull request merged**